Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java

```diff
@@ -505,45 +505,64 @@ public boolean isActive() throws IOException {
-        App yarnApp = this.getSparkJobYarnApplication(this.getConnectUri(), applicationId);
+    public Observable<SimpleImmutableEntry<SparkBatchJobState, String>> getJobDoneObservable() {
+        return Observable.create((Subscriber<? super SimpleImmutableEntry<SparkBatchJobState, String>> ob) -> {
+            try {
+                boolean isJobActive = true;
+                boolean isLogAggregateDone = false;
+                SparkBatchJobState state = SparkBatchJobState.NOT_STARTED;
+                String applicationId = null;
+                String diagnostics = "";
+
+                while (isJobActive || !isLogAggregateDone) {
+                    if (isJobActive) {
+                        HttpResponse httpResponse = this.getSubmission().getBatchSparkJobStatus(
+                                this.getConnectUri().toString(), batchId);
+
+                        if (httpResponse.getCode() >= 200 && httpResponse.getCode() < 300) {
+                            SparkSubmitResponse jobResp = ObjectConvertUtils.convertJsonToObject(
+                                    httpResponse.getMessage(), SparkSubmitResponse.class)
+                                    .orElseThrow(() -> new UnknownServiceException(
+                                            "Bad spark job response: " + httpResponse.getMessage()));
+
+                            state = SparkBatchJobState.valueOf(jobResp.getState().toUpperCase());
+
+                            isJobActive = !state.isJobDone();
+                            applicationId = jobResp.getAppId();
+                        }
+                    } else {
+                        App yarnApp = this.getSparkJobYarnApplication(this.getConnectUri(), applicationId);
+                        diagnostics = yarnApp.getDiagnostics();
+
+                        switch (yarnApp.getLogAggregationStatus().toUpperCase()) {
+                            case "SUCCEEDED":
+                            case "FAILED":
+                                isLogAggregateDone = true;
+                                break;
+                            case "DISABLED":
+                            case "NOT_START":
+                            case "RUNNING":
+                            case "RUNNING_WITH_FAILURE":
+                            case "TIME_OUT":
+                            default:
+                                isLogAggregateDone = false;
+                        }
+                    }
 
-        switch (yarnApp.getLogAggregationStatus().toUpperCase()) {
-            case "SUCCEEDED":
-                return true;
-            case "DISABLED":
-            case "NOT_START":
-            case "RUNNING":
-            case "RUNNING_WITH_FAILURE":
-            case "FAILED":
-            case "TIME_OUT":
-            default:
-                return false;
-        }
-    }
+                    // Retry interval
+                    sleep(1000);
```

Comment:
>sleep(1000); [](start = 20, length = 12)

in another thread for safety since it's interval querying?

Reply:
Yes, the caller uses `Schedulers.io()`